mautic/mautic
# Password  complexity check

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? |
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | https://github.com/bjeavons/zxcvbn-php
| Issues addressed (#s or URLs) | /
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is an alternative to https://github.com/mautic/mautic/pull/12812 and the community must decide which PR to merge. This PR is better in the way that:
1. It doesn't check strict rules of what characters the password contains, but the complexity combination together with dictionary check for frequent words.
2. It forces current users to reset passwords on login so all users will be forced to use more secure passwords.
3. Works from all ways where to create user (installation, API, UI)

We must ensure that admin users are not allowed to use simple passwords.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a user in admin panel with a very simple password like `11111111`. You will be allowed to do it.

#### Steps to test this PR:
1. Try to login with a user that has a simple password. You will be redirected to a password reset page with a message that the you have to change the password.
2. Make sure the email worker is running with `bin/console messenger:consume email` command
3. Change the password to a strong one and login to the admin panel.
4. Try to create a user with a simple password. You shouldn't be allowed to do it. There should be a validation error that your password is too weak.
5. Test also that the password weakness is also checked when creating the admin user during installation
